cmd/restic/cmd_rebuild_index.go

```diff
@@ -44,58 +52,109 @@ func runRebuildIndex(gopts GlobalOptions) error {
 		return err
 	}
 
-	ctx, cancel := context.WithCancel(gopts.ctx)
-	defer cancel()
-	return rebuildIndex(ctx, repo, restic.NewIDSet())
+	return rebuildIndex(opts, gopts, repo, restic.NewIDSet())
 }
 
-func rebuildIndex(ctx context.Context, repo restic.Repository, ignorePacks restic.IDSet) error {
-	Verbosef("counting files in repo\n")
+func rebuildIndex(opts RebuildIndexOptions, gopts GlobalOptions, repo *repository.Repository, ignorePacks restic.IDSet) error {
+	ctx := gopts.ctx
 
-	var packs uint64
-	err := repo.List(ctx, restic.PackFile, func(restic.ID, int64) error {
-		packs++
+	var obsolete restic.IDs
+	err := repo.List(ctx, restic.IndexFile, func(id restic.ID, size int64) error {
+		obsolete = append(obsolete, id)
 		return nil
 	})
 	if err != nil {
 		return err
 	}
 
-	bar := newProgressMax(!globalOptions.Quiet, packs-uint64(len(ignorePacks)), "packs")
-	idx, invalidFiles, err := index.New(ctx, repo, ignorePacks, bar)
-	if err != nil {
-		return err
-	}
+	packs := restic.NewIDSet()
+	packsWronglyIndexed := restic.NewIDSet()
+	totalBlobs := uint(0)
+
+	if opts.ReadAllPacks {
+		Verbosef("finding pack files in repo\n")
+		err = repo.List(ctx, restic.PackFile, func(id restic.ID, size int64) error {
+			packs.Insert(id)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	} else {
+		Verbosef("load indexes\n")
+		err = repo.LoadIndex(gopts.ctx)
+		if err != nil {
+			return err
+		}
+
+		type PackInfo struct {
+			size  uint64
+			blobs uint
+		}
+		packInfo := make(map[restic.ID]PackInfo)
+
+		Verbosef("get pack files to read\n")
+		// iterate over all blobs in index
+		for blob := range repo.Index().Each(ctx) {
+			pi, ok := packInfo[blob.PackID]
+			if !ok {
+				pi.size = pack.HeaderSize
+			}
+			pi.size += uint64(pack.PackedSizeOfBlob(blob.Length))
+			pi.blobs++
+			// update packSizeFromIndex
+			packInfo[blob.PackID] = pi
+		}
 
-	if globalOptions.verbosity >= 2 {
-		for _, id := range invalidFiles {
-			Printf("skipped incomplete pack file: %v\n", id)
+		err := repo.List(ctx, restic.PackFile, func(id restic.ID, packSize int64) error {
+			pi, ok := packInfo[id]
+			if !ok || pi.size != uint64(packSize) {
+				// Pack was not referenced in index or size does not match
+				packs.Insert(id)
+			} else {
+				totalBlobs += pi.blobs
+			}
+			delete(packInfo, id)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+		// process pack files that are referenced in the index but do not exist
+		for id := range packInfo {
+			packsWronglyIndexed.Insert(id)
```

Comment:
`missingPacks` would be a more accurate description.

Reply:
You are right, that is the better name - is changed.